Task1

AlexandrMiloradov opened the pull request:
Well done !

AlexandrMiloradov committed:
task1

AlexandrMiloradov committed:
task1

AlexandrMiloradov commented on src/main/java/start.java:
```diff
@@ -1,9 +1,15 @@
-    public static void main(String args[]){
-
+    public static void main(String args[]) throws IOException {
+        weather weatherEkb = new weather(56.8519, 60.6122, true, (short) 3);
```
Получаю экземпляры класса с погодой на 3 города

AlexandrMiloradov commented on src/main/java/factWeather.java:
```diff
@@ -0,0 +1,51 @@
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class factWeather extends infoWeather {
```
Фактическая погода - присылается в ответ на запрос к api

krutakod replied:
- [x] Плохие имена всех классов (CamelName) и лучше использовать реальные английские слова, например, вместо "FactWeather" лучше использовать "ActualWeather".

AlexandrMiloradov commented on src/main/java/forecast.java:
```diff
@@ -0,0 +1,83 @@
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class forecast  {
```
Объект прогноз - присылается в ответ на запрос. Содержит в себе прогноз на день в разрезе частей суток

krutakod replied:
- [x] Нужно будет во всех местах проекта исправить нейминг

AlexandrMiloradov commented on src/main/java/forecast.java:
```diff
@@ -0,0 +1,83 @@
+    private String sunset;
+    private short moon_code;
+    private String moon_text;
+    private partsOfWeather parts;
```
содержит прогноз внутри на время суток. Выделен в класс так как имеет разный состав полей и в коллекцию пропихнуть нельзя

AlexandrMiloradov commented on src/main/java/infoPlace.java:
```diff
@@ -0,0 +1,60 @@
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class infoPlace {
```
информация о месте прогноза

AlexandrMiloradov commented on src/main/java/infoWeather.java:
```diff
@@ -0,0 +1,200 @@
+import com.fasterxml.jackson.annotation.*;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class infoWeather {
```
класс обертка для прогнозов. содержит поля используемые везде для того чтобы не дублировать поля

AlexandrMiloradov commented on src/main/java/partsOfWeather.java:
```diff
@@ -0,0 +1,63 @@
+import com.fasterxml.jackson.annotation.*;
+
+public class partsOfWeather {
```
класс содержащий внутри прогноз на части суток

AlexandrMiloradov commented on src/main/java/timeZone.java:
```diff
@@ -0,0 +1,41 @@
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class timeZone {
```
класс определяющий часовую зону

AlexandrMiloradov commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+import java.util.List;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class weather {
```
общий класс для прогноза

krutakod commented on src/main/java/factWeather.java:
```diff
@@ -0,0 +1,51 @@
+
+    private double temp;
+    private String season;
+    private double obs_time;
```
- [x] Переменные нужно переименовать по стилю "CamelName" или же "ObsTime"

krutakod commented on src/main/java/factWeather.java:
```diff
@@ -0,0 +1,51 @@
+        return phenom_condition;
+    }
+
+    public void setPhenom_condition(String phenom_condition) {
```
- [x] Наименование set и get методов по тому же принципу должны быть

krutakod commented on src/main/java/infoPlace.java:
```diff
@@ -0,0 +1,60 @@
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Для чего использовал эту аннотацию? Здесь просто я сам интересуюсь. Не говорю, что неправильно.

AlexandrMiloradov replied:
В ответе от яндекса есть поля которые отсутствуют в классе. Поэтому использую эту аннотацию

krutakod commented on src/main/java/infoWeather.java:
```diff
@@ -0,0 +1,200 @@
+    @JsonIgnore
+    private double temp;
+    private double feels_like;
+    @JsonIgnore
```
- [x] Почему часть полей игнорится? Точнее, для чего была добавлена аннотация @JsonIgnore. Она считается плохим стилем, потому что если тебе нужно игнорировать какие-то поля, то это указывает на проблему архитектуры (тип класс описывает JSON, значит, все поля у нас не игнорируются), которую нужно решить и не использовать @JsonIgnore

AlexandrMiloradov replied:
Игнорирую поля чтобы в файл выгружалось только то что нужно в задаче

krutakod replied:
Значит, надо пересмотреть классы, чтобы уйти от @JsonIgnore

krutakod commented on src/main/java/start.java:
```diff
@@ -1,9 +1,15 @@
+        weather weatherEkb = new weather(56.8519, 60.6122, true, (short) 3);
+        weather weatherNY = new weather(40.7143, -74.006, true, (short) 3);
+        weather weatherLon = new weather(51.5085, 0.12574, true, (short) 3);
+        List<weather> list = Arrays.asList(weatherEkb, weatherNY,weatherLon);
```
- [x] Не используй Arrays.asList(). Нужно будет убрать. Если интересно почему нельзя использовать его, то вот https://telegra.ph/Pravilnoe-preobrazovanie-massiva-v-ArrayList-12-26, но решение тебе нужно будет найти самостоятельно (из статьи не подойдёт)

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+
+    }
+
+    public weather(double lat, double lon, boolean extra, short days) throws IOException {
```
- [x] В конструкторе может быть максимум инициализация полей. Вся любая другая логика должна быть вынесет в другие методы и классы.

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+        this.forecasts = forecasts;
+    }
+
+    private void setTemp_avg(){
```
В методе set не должно быть сложной логики. Его максимум - простой if. Хотя даже такое считается плохим тоном, так как валидации данных реализуют по-другому. Напрмер, после создания объекта, когда метод set-отработал. Но это все лирика. В будущем увидишь. 

- [x] Короче, нужно эту логику вынести из set-метода.

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+    private void setTemp_avg(){
+        int i = 0;
+        double temp_avg = 0;
+        for (forecast forecast:forecasts) {
```
Вот тебе пример простой пример плохо читаемого кода. Класс называется forecast и переменная называется forecast. Других разработчиков это смутит. Поэтому классы называются всегда с большой буквы, чтобы их было видно в коде. А названия переменных и методов всегда начинают в маленькой буквы. Ну и используем CamelNames или camelNames соответственно.

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+        for (forecast forecast:forecasts) {
+           infoWeather day_forecast = forecast.getParts().getDay();
+           temp_avg = temp_avg+day_forecast.getTemp_avg();
+           i++;
```
- [x] Если ты считаешь количество итераций, то проще будет воспользоваться простым for, чтобы не изобретать велосипеды с forEach + счётчик, отдельно вынесенный. Плюс так легче будет читать код

 forEach стоит использовать, когда не нужно считать количество итераций.

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+        double temp_avg = 0;
+        for (forecast forecast:forecasts) {
+           infoWeather day_forecast = forecast.getParts().getDay();
+           temp_avg = temp_avg+day_forecast.getTemp_avg();
```
Можно упростить оператором +=

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+        String jsonString = "{"+dm+"cities"+dm+":[";
+        ObjectMapper jsonMapper = new ObjectMapper();
+        for (weather weather:weatherObjects) {
+            jsonString = jsonString + jsonMapper.writeValueAsString(weather)+",";
```
- [x] Стоит упростить с помощью +=

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+    }
+    public static void writeRequest(List<weather> weatherObjects, String catalog) throws IOException {
+        final char dm = (char) 34;
+        String jsonString = "{"+dm+"cities"+dm+":[";
```
Не очень понимаю, зачем нужна такая конструкция с большим количеством соединений строк. Но выглядит как костыль. Думаю, весь этот код - это попытка в один JSON массив положить десериализованные объекты и записать потом это все в файл. Но для любых операций сериализации/десериализации мы используем исключительно jackson или другие библиотеки для работы с JSON. Не изобретай велосипеды, они медленнее будут работать, а ещё заставляют тебя думать, как решить уже решенные задачи. Думать вредно. 

- [ ] Попробуй сформировать JSON необходимый с помощью jackson.

AlexandrMiloradov replied:
Здесь я формирую json строку. В этом методе в качестве аргументов передается лист с погодами для городов. Соотвественно я в json строку добавляю массив cities - далее идет погода. Сделал таким образом так как нужны кавычки. Я не нашел альтернативного исполнения - поэтому сделал так

krutakod replied:
Ага, альтернативный способ существует для формирования JSON-строки из классов, списков и т.п. Поэтому нужно будет найти, как реализовать. Иначе руками было бы очень сложно формировать большие JSON-строки.

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+        jsonString = jsonString.substring(0, jsonString.length()-1) + "]}";
+        FileWriter fileWriter = new FileWriter(catalog);
+        fileWriter.write(jsonString);
+        fileWriter.close();
```
А что будет, если при попытке записи возникнет исключение по какой-то причине? Тогда поток записи не закроется, а любой незакрытый поток чткния/записи - это полный булщит. Советую загуглить и внимательно ознакомиться с этим. Просто мы, как разработчики, должны гарантировать закрытие всех потоков всегда. Исключений нет. Любое исключение в этом случае - непредсказуемая ситуация. ъ

- [x] Поэтому обработай исключение IO в этом методе и гарантирую закрытие потоков.

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+            this.temp_avg = temp_avg/i;
+        }
+    }
+    public static void writeRequest(List<weather> weatherObjects, String catalog) throws IOException {
```
- [ ] Получается, если нужно будет этот метод использовать, то ты передашь всю ответственность за исключение разработчикам, которые используют твой метод. Об этом говорит throws IOException. Получается они будут по всей программе плодить блоки try-catch. Лучше лишить их этого кайфа и обработать исключение внутри этого метода. Тем более, в данном случае нет смысла давать свободу в обработки исключений другим разработчикам, которые будут использовать этот метод, так как здесь все однозначно.

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+    public weather(double lat, double lon, boolean extra, short days) throws IOException {
+
+        this.days = days;
+        String urlRequest = weather.url+"lat="+lat+"&"+"lon="+lon+"&"+"extra="+Boolean.toString(extra)+"&"+"limit="+days;
```
- [x] Выглядит костыльно формирование url. Попробуй использовать String.format() - надеюсь, что правильно помню. А ещё можно попробовать использовать паттерн Строитель вместо String.format() для формирования url запроса для тренировки реализации паттернов.

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+    public weather(double lat, double lon, boolean extra, short days) throws IOException {
+
+        this.days = days;
+        String urlRequest = weather.url+"lat="+lat+"&"+"lon="+lon+"&"+"extra="+Boolean.toString(extra)+"&"+"limit="+days;
```
- [ ] weather.url, насколько понимаю, это базовый url. Его нужно вынести в отдельную константу, используемую при формировании url. Он всегда должен быть один и тот же, а константой мы это гарантируем.

AlexandrMiloradov replied:
url вынесен в переменную в классе с модификатором static

krutakod replied:
static - это не константа. Тем более, лучше для констант создавать отдельные классы/интерфейсы/перечисления(enum).

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+        this.days = days;
+        String urlRequest = weather.url+"lat="+lat+"&"+"lon="+lon+"&"+"extra="+Boolean.toString(extra)+"&"+"limit="+days;
+        final URL url = new URL(urlRequest);
+        final HttpURLConnection con = (HttpURLConnection) url.openConnection();
```
- [x] Отправка запросов у тебя явно будет где-то ещё в программе. Поэтому было бы прикольно написать универсальный метод, который бы выполнял запросы и возвращал ответ на них, чтобы во всей программе было меньше копипаста одного и того же кода + раз выполнение запроса будет только в одном месте, то легче будет контролировать исключения.

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+        final URL url = new URL(urlRequest);
+        final HttpURLConnection con = (HttpURLConnection) url.openConnection();
+        con.setRequestMethod("GET");
+        con.setRequestProperty("X-Yandex-API-Key", weather.apiKey);
```
- [x] Ключ API должен стать тоже константой, как и выражение X-Yandex-API-Key

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+        final HttpURLConnection con = (HttpURLConnection) url.openConnection();
+        con.setRequestMethod("GET");
+        con.setRequestProperty("X-Yandex-API-Key", weather.apiKey);
+        try (final BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()))) {
```
- [x] Чтение ответа - это тоже будет отдельная и универсальная функция, потому что нет смысла городить один и тот же код по всей программе.

krutakod commented on src/main/java/weather.java:
```diff
@@ -0,0 +1,137 @@
+                setTemp_avg();
+            }
+        } catch (final Exception ex) {
+            ex.printStackTrace();
```
- [x] При обработке исключений нужно избегать использование Exception, так как это общий класс для любых исключений. И исключения мы должны классифицировать, чтобы понимать, что и где пошло не так. Нужно уйти от использования общего класса Exception к частным и конкретным.

krutakod replied:
- [x] У тебя нигде нет обработки отрицательного ответа от яндекс API. Например, оно может дать тебе 404 ошибку или 403, а может и 500 и кучу других кодов. Нужно ловить отрицательные ответы API и выбрасывать соответствующие исключения с описанием проблемы. Для обработки таких исключений создай свой собственный класс исключений.

AlexandrMiloradov committed:
v2

AlexandrMiloradov committed:
v2

AlexandrMiloradov committed:
v3

AlexandrMiloradov committed:
v3 weather exception

krutakod commented on src/main/java/InfoWeather.java:
```diff
@@ -0,0 +1,73 @@
+        this.tempMin = tempMin;
+    }
+
+    @JsonGetter("tempMax")
```
Что такое JsonSetter? И зачем он применяется у всех полей класса?

krutakod replied:
Думаю, это легко заменяется JsonProperty. Если что, поправь, потому что мне тоже самому интересно

AlexandrMiloradov replied:
Наименования полей разные в классе и ответе яндекса поэтому я использую эту аннотацию чтобы прямо указывать jackson что является сеттером и геттером

krutakod replied:
Давай уйдём в сторону @JsonProperty, так как она лучше читаема

krutakod commented on src/main/java/Weather.java:
```diff
@@ -0,0 +1,156 @@
+    private String nowDt;
+    private InfoPlace info;
+    private List<Forecast> forecasts = new ArrayList<>();
+    private final static String nameApiKey = "X-Yandex-API-Key";
```
Это не относится к модели Погода. Если у тебя есть класс-модель, описывающий какую-то сущность, тогда он должен быть максимально простым. Модель - это POJO java-класс. Можешь загуглить. Но вот неплохая статья с примером:
https://devcolibri.com/unit/урок-8-создание-pojo-объекта-user-работа-с-view-из-java-2/

krutakod commented on src/main/java/Weather.java:
```diff
@@ -0,0 +1,156 @@
+    private void setTempAvg(){
+        int i = 0;
+        double temp_avg = 0;
+        for (i = 0; i < forecasts.size(); i ++) {
```
в set-рах только инициализация может быть. всю логику нужно выносить из них.

krutakod commented on src/main/java/Weather.java:
```diff
@@ -0,0 +1,156 @@
+        }
+    }
+
+    public void getWeatherFromApi() throws IOException, APIYandexExceptions {
```
Это не должно быть в POJO-классе модели.

krutakod commented on src/main/java/Weather.java:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    public void getWeatherFromApi() throws IOException, APIYandexExceptions {
+        UrlConnector urlConnector = new UrlConnector(getUrlStringForApiRequest());
```
- [x] Необходимо вынести этот getWeatherFromApi() в отдельный класс и метод.

Назови класс UrlRequest. Пусть в нём будет метод sendRequest(), вот он будет отправлять ЛЮБЫЕ get-запросы. Он будет ключевым и универсальным, чтобы мы могли использовать его для отправки get-запросов на любые API. Пусть метод на вход принимает ссылку на объект класса RequestData. И данный метод должен возвращать ответ на запрос.

RequestData - это класс, который будет у тебя описывать заголовки url (String), (headers, map-коллекция), список параметров (params, map-коллекция) и тело (body, String - нужен для post-запросов) для любых типов запросов. Как будет работать этот RequestData: 
1) это реализация паттерна builder (строитель)
2) по-умолчанию для каждого объекта создаются пустые коллекции
Смысл в том, что этот объект данного класса будет описывать все данные для любого типа запросов. Например, в классе будет метод header(String headerName, String headerValue): он добавляет в headers переданный заголовок. Подобные методы будут и для других полей объекта.

krutakod commented on src/main/java/Weather.java:
```diff
@@ -0,0 +1,156 @@
+        urlConnector.openConnection();
+        urlConnector.setTheHeadersToUrl(getNameApiKey(), getApiKey());
+        StringReader stringReader = urlConnector.getAnswerFromUrl();
+        if (stringReader == null){
```
- [x] А если будет не null, а какая-то ошибка? Необходимо посмотреть в API яндекса, какие коды ошибок может отправлять api клиенту, чтобы этот список твоя программа обрабатывала и вызывала исключение с полезной информацией о том, что пошло не так.

krutakod commented on src/main/java/Weather.java:
```diff
@@ -0,0 +1,156 @@
+        urlConnector.setTheHeadersToUrl(getNameApiKey(), getApiKey());
+        StringReader stringReader = urlConnector.getAnswerFromUrl();
+        if (stringReader == null){
+            throw new APIYandexExceptions("Wrong url address");
```
- [x] Очень сомневаюсь, что если url будет неправильным, то программа вообще дойдёт до этой строчки. Но протестируй, мне тоже интересно

krutakod commented on src/main/java/Weather.java:
```diff
@@ -0,0 +1,156 @@
+        }
+    }
+
+    public void getWeatherFromJsonString(StringReader stringReader) throws IOException {
```
- [x] Создай класс-сервис. Будет называться YandexWeatherService. Он уже будет отвечать за отправку запроса в API яндекса в помощью UrlRequest. И в нём будет, например, метод getWeatherForPlace(). Пусть метод принимает ссылку на объект FilterYandexWeatherData (широта, долгота, количество дней и пр. параметры, если нужны). И вот этот метод пусть возвращает список объектов исправленного класса Weather (ну или просто объект Weather, это уже будет зависеть от тебя, но, наверное, логичнее список).

И вот в YandexWeatherService будет getWeathersFromJsonString(), private метод, который тупо будет парсить ответ от яндекса (private, так как зачем этот метод использовать где-то ещё, если он только парсит ответ яндекса). В нём нужно вызывать исключения, если яндекс возвращает ошибки, а если  всё ок, то пусть возвращает список объектов Weather.

krutakod commented on src/main/java/Weather.java:
```diff
@@ -0,0 +1,156 @@
+        jsonString = jsonString.substring(0, jsonString.length()-1) + "]}";
+        FileWriter fileWriter = new FileWriter(catalog);
+        fileWriter.write(jsonString);
+        fileWriter.close();
```
- [x] небезопасное закрытие потоков чтения/записи.  если где-то выше будет вызвано исключение, то поток не закроется и ресурс будет занят. нам оно не надо

krutakod commented on src/main/java/UrlConnector.java:
```diff
@@ -0,0 +1,60 @@
+            con = (HttpURLConnection) url.openConnection();
+            con.setRequestMethod("GET");
+        } catch (IOException ex) {
+            System.out.println("Wrong url address");
```
- [x] Очень плохо, так как ты не выкидываешь в консоль информацию об исключении. А ещё и заменяешь любые исключения надписью "неверный url", хотя, например, сервер может быть недоступен просто из-за отключённого интернета на твоём компьютере или по сотне других причин. А у тебя во всех случаях будет просто указываться, что "Неверный URL". Надо исправить.

krutakod commented on src/main/java/UrlConnector.java:
```diff
@@ -0,0 +1,60 @@
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+public class UrlConnector {
```
Весь этот класс по сути станет UrlRequest, который описал ниже.

krutakod commented on src/main/java/APIYandexExceptions.java:
```diff
@@ -0,0 +1,8 @@
+import java.io.IOException;
+
+public class APIYandexExceptions extends IOException {
```
- [x] Почему от IOException наследуешься? Подумай, от какого класса-исключения лучше наследоваться (смотри иерархию исключений). Например, если мы укажем неверные данные ключа для API это же никак не относится к потокам чтения/записи.

krutakod commented on src/main/java/Weather.java:
```diff
@@ -0,0 +1,156 @@
+        weather = null;
+    }
+
+    public static void writeRequest(Weather weather, String catalog) throws IOException {
```
- [x] Я бы вынес в отдельный класс и метод, который бы принимал строку и тупо записывал бы её в указанный файл. Чтобы потом в любом месте программы записывать строку в файл можно было.

AlexandrMiloradov committed:
v3 weather exception
Добавил новые классы, переписал класс запросов к серверу, вынес код парсера в отдельный класс

AlexandrMiloradov committed:
v3 weather exception
Добавил новые классы, переписал класс запросов к серверу, вынес код парсера в отдельный класс

AlexandrMiloradov committed:
v4 weather
Добавил новые классы, переписал класс запросов к серверу, вынес код парсера в отдельный класс

krutakod commented on src/main/java/WeatherService.java:
```diff
@@ -0,0 +1,42 @@
+
+public class WeatherService {
+
+    public static void setTempAvg(Weather weather){
```
- [x] Что делает данный метод? И почему он находится в сервисе? Если он просто считает среднюю температуру, тогда необходимо изменить название и сам метод. Например, getAvgTemp(List<Forecast> forecasts), тогда он принимать forecasts  и будет возвращать значение средней температуры. 

Почему так: 
1) когда ты используешь в названиях приставку set - это всегда намёк на какой-то метод setter в классе модели, хотя в этом случае это вообще не так. если другой прогер такое увидит, то он не поймёт поо назвнаию, что делает этот метод и пойдёт смотреть его код, а нам это не надо. 
2) когда ты делаешь метод, то лучше (когда это просто сделать) отправлять в метод только необходимую для его работы информацию. Например, для расчёта средней погоды нужен лишь список List<Forecast> forecasts, получается, что нет смысла отправлять всю модельку Weather. А ещё почему это прикольно: вдруг у тебя появится WeatherV2 какой-нибудь, тогда нужно будет решать проблему, что метод принимает только Weather.
3) А вдруг нам нужно просто среднее значение температуры без записи в модель. Нам всё равно на всю модель, а важно лишь среднюю температуру использовать в дальнейшем почему-то.

Теперь у тебя будет следующая логика кода: прогер вызывает метод getAvgTemp(List<Forecast> forecasts), отправляет в него список forecasts, получает среднюю температуру и с помощью set-метода модели Weather инициализирует необходимую ему модель. И, как видишь, такой код более удобно читать всем.

krutakod replied:
Прикольно, что этот метод реализовал в общем сервисе Weather без привязки к конкретной погоде (например, яндекс)

krutakod commented on src/main/java/WeatherService.java:
```diff
@@ -0,0 +1,42 @@
+
+    }
+
+    public static void writeRequest(Weather weather, String catalog) throws IOException {
```
Хороший метод без всякого лишнего мусора, отлично, что void.

krutakod commented on src/main/java/WeatherService.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    public static void writeRequest(Weather weather, String catalog) throws IOException {
+        FileWriter fileWriter = new FileWriter(catalog);
```
try с ресурсами не работает с FileWriter?

krutakod commented on src/main/java/YandexWeatherService.java:
```diff
@@ -0,0 +1,59 @@
+    private final static String url = "https://api.weather.yandex.ru/v2/forecast?";
+
+    private static Weather getWeatherFromJsonString(StringReader stringReader) throws IOException {
+
```
- [x] Нет смысла в лишних переносах строк в маленьких методах, так как у них простая и очевидная логика, которую не надо разделать переносами строк.

krutakod commented on src/main/java/YandexWeatherService.java:
```diff
@@ -0,0 +1,59 @@
+        catch (APIYandexExceptions ex){
+            System.out.println(ex.getMessage());
+        }
+        catch (IOException ex){
```
- [x] Эти два try-catch можно объединить в один и использовать множественный catch

krutakod commented on src/main/java/UrlConnector.java:
```diff
@@ -0,0 +1,87 @@
+                stringReader = new StringReader(inputLine);
+            }
+        } else {
+           throw new UrlConnectorException("Server send request with "+"["+Integer.toString(responseCode)+"] code");
```
- [x] В это случае (как и в подобных) лучше использовать String.format(), будет более читаемый код, особенно когда много всяких аргументов

krutakod commented on src/main/java/UrlConnector.java:
```diff
@@ -0,0 +1,87 @@
+           throw new UrlConnectorException("Server send request with "+"["+Integer.toString(responseCode)+"] code");
+        }
+
+        return stringReader;
```
молодец, что только один выход из метода реализовал

krutakod commented on src/main/java/UrlConnector.java:
```diff
@@ -0,0 +1,87 @@
+        }
+    }
+
+    private StringReader getAnswerFromUrl() throws IOException, UrlConnectorException {
```
- [ ] Этот метод всё ещё нуждается в том, что его нужно вынести в другой класс-сервис.

krutakod commented on src/main/java/UrlConnector.java:
```diff
@@ -0,0 +1,87 @@
+        int responseCode = con.getResponseCode();
+
+        if (responseCode == 200){
+            final BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()));
```
- [ ] модификатор final в данном случае точно требуется?

krutakod commented on src/main/java/YandexWeatherService.java:
```diff
@@ -0,0 +1,59 @@
+        UrlConnector urlConnector = new UrlConnector(url);
+
+        Map<String, String> params = new LinkedHashMap<>();
+        params.put("lat", Double.toString(filterYandexWeatherData.getLat()));
```
Давай тогда пока что сделаешь всё, что написал, а потом мы ещё вот эти всякие параметры сделаем с тобой более красивые, чтобы паттерн Строитель появился для красоты и для освоения материала.

krutakod commented on src/main/java/UrlConnector.java:
```diff
@@ -0,0 +1,87 @@
+        this.urlString = urlString;
+    }
+
+    private void setTheHeadersToUrl(Map<String, String> headers){
```
Это тоже нужно будет вынести, но я скажу позже, как лучше реализовать, чтобы сейчас тебя с других изменениями не сбивать.

AlexandrMiloradov committed:
Обработал последний комментарии

AlexandrMiloradov committed:
Обработал последний комментарии
